[Lee Hong Sheng] iP

hongshenggg opened the pull request:
# DukePro
>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. Download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html)
2. Double-click it
3. Add your tasks
4. Let it manage it for you :wink:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (Coming soon)
- [ ] Reminders (Coming soon)

<hr>

Code block
```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

hongshenggg committed:
Completed Level-1

hongshenggg committed:
Duke is now able to add to list of tasks

hongshenggg committed:
Completed level 2

hongshenggg committed:
Completed Level 3

hongshenggg committed:
Completed level 4

hongshenggg committed:
Completed A-TextUiTesting

hongshenggg committed:
Complete level 5

hongshenggg committed:
Complete Level 6

hongshenggg committed:
Update task types to use enums

hongshenggg committed:
Modify switch-case statements to follow coding convention

hongshenggg committed:
Complete branch-Level-7

hongshenggg committed:
Use LocalDate for Deadline and Event tasks

hongshenggg committed:
Merge branch 'branch-Level-7'

hongshenggg committed:
Merge branch-Level-8

hongshenggg committed:
Refactor the code to extract out related classes

hongshenggg committed:
Refactor codebase into suitable packages

hongshenggg committed:
Merge branch 'add-gradle-support'

hongshenggg committed:
Add JUnit tests for Parser and Tasklist

hongshenggg committed:
Complete A-Jar

hongshenggg committed:
Add JavaDoc comments to the source code

hongshenggg committed:
Refactor code to better comply with coding standard

hongshenggg committed:
Add the feature to find a task based on a keyword

hongshenggg committed:
Merge branch 'branch-A-CodingStandard'

hongshenggg committed:
Merge branch 'branch-Level-9

hongshenggg committed:
Add GUI to Duke

Acerizm commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,170 @@
+                    + matchingTasks;
+        }
+    }
+}
```
Overall codebase is very neat! Keep it up! I would suggest that each method inside Duke.java can be extracted out as commands so that Duke can only be responsible for mainly running the "run" command

Acerizm commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,157 @@
+    private void retrieveDates(TaskTypes type, String input)
+            throws DukeException {
+        String[] temp;
+        switch (type) {
+        case DEADLINE:
+            if (!input.contains("/by")) {
+                throw new DukeException("Please provide a deadline using /by");
+            }
+            temp = input.split(" /by ");
+            if (temp.length < 2) {
+                throw new DukeException("Please provide a valid date.");
+            }
+            try {
+                this.dates[0] = LocalDate.parse(temp[1]);
+            } catch (DateTimeParseException e) {
+                throw new DukeException(
+                        "Please provide a valid date in the following "
+                                + "format: YYYY-MM-DD.");
+            }
+            break;
+        case EVENT:
+            if (!input.contains("/from") || !input.contains("/to")) {
+                throw new DukeException(
+                        "Please provide a start date and end date "
+                                + "using /from and /to respectively.");
+            }
+            temp = input.split(" /from ");
+            String[] dates = temp[1].split(" /to ");
+            if (dates.length < 2) {
+                throw new DukeException("Please provide a valid start and end date.");
+            }
+            try {
+                this.dates[0] = LocalDate.parse(dates[0]);
+                this.dates[1] = LocalDate.parse(dates[1]);
+            } catch (DateTimeParseException e) {
+                throw new DukeException(
+                        "Please provide the dates in the following "
+                                + "format: YYYY-MM-DD.");
+            }
+            break;
```
If you're keen to explore you could try writing DukeException to automatically print specific messages based on the task and error in DukeException.java. Currently, DukeException behaviour is no different than a normal Exception class since you could also pass the same messages to an Exception object and throw it.

Overall the current idea of throwing DukeExceptions is good too!

Acerizm commented on src/main/java/duke/storage/TaskStorage.java:
```diff
@@ -0,0 +1,183 @@
+            tasks.add(t);
+        }
+        return tasks;
+    }
```
I like the way you handled loading the data! Good work!

Acerizm reviewed (commented):
Excellent work and keep it up! Left some potential comments/improvement ideas for you to try out!

hongshenggg committed:
Add GUI to Duke

Duke is a CLI program that can be unintuitive to users
that are not familiar with CLI applications.

A GUI provides users with a more familiar and intuitive
environment to interact with and utilise Duke.

Let's implement a GUI for Duke using JavaFX libraries.

cetigerlily commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,170 @@
+        this.isActive = true;
+    }
+
+    public String getResponse(String input) {
```
it's great to see that you have added javadocs to a lot of your methods, maybe when you have time, you could add them for these methods too?

cetigerlily reviewed (commented):
your project looks great!

hongshenggg committed:
Use assert feature to document assumptions in the code

hongshenggg committed:
Refactor to improve code quality

hongshenggg committed:
Use streams

hongshenggg committed:
Merge pull request #1 from hongshenggg/branch-A-Assertions

Use assert feature to document assumptions in the code

hongshenggg committed:
Merge branch 'master' into branch-A-CodeQuality

hongshenggg committed:
Merge branch 'master' into branch-A-Streams

hongshenggg committed:
Merge pull request #2 from hongshenggg/branch-A-CodeQuality

Refactor to improve code quality

hongshenggg committed:
Merge branch 'master' into branch-A-Streams

hongshenggg committed:
Merge pull request #3 from hongshenggg/branch-A-Streams

Use streams

hongshenggg committed:
Allow Duke to detect duplicates when adding a new task to the list

hongshenggg committed:
Merge branch 'branch-C-DetectDuplicates'

hongshenggg committed:
Fix the bug where marking a task results in data file to be cleared

hongshenggg committed:
Add a user guide and Ui screenshot

hongshenggg committed:
Fix javafx warning
